src/docs/arch-concurrency.dox

```diff
@@ -0,0 +1,107 @@
+the WiredTiger generations system. In general when a reader thread accesses an object, that could be
+concurrently freed, it enters a "generation". After entering that generation it accesses the object.
+A generation is a monotonically increasing integer which denotes times at which a given object must
+exist.
+
+Writer threads are then able to replace that object but cannot free it until the reader thread has
+left or released the generation. Instead the writer thread adds the object to a "stash" and it marks
+the stashed object with the generation indicating the point at which it was added to the stash. Once
+all reader in the system do not have a corresponding generation which is older than or equal to the
+stashed object's generation it can be freed.
+
+@section atomics Atomics
+
+As previously mentioned WiredTiger already expects loads and stores to specific sizes of points
+to be atomic and prevent tearing. WiredTiger also relies on atomic RMW operations, these include:
+- Compare and Swap
+- Fetch add / sub
+- Add / sub fetch
+
+Currently WiredTiger is undergoing a transition where all shared memory variable access will be
+wrapped by an atomic function, this will enable TSan to run on the code base. To facilitate this
+transition the atomic load and store macros have been added to WiredTiger, these macros placate
+TSan. To prevent them from introducing a performance regression they utilize relaxed consistency.
+
+@section memory-barriers Memory Barriers
+
+WiredTiger utilizes memory barriers in a number of places to manage CPU instruction reordering.
+Specifically CPUs are allowed to reorder load and store instructions. If this occurs when two
+threads are reading or writing to the same shared memory locations it can result in those threads
+seeing what would otherwise be impossible values.
+
+WiredTiger defines barriers by describing which kind of reordering they prevent, for example a
+LoadLoad barrier prevents load instructions before the barrier from being reordered with load
+instructions after the barrier. Using this definition we can define 4 types of reorderings,
+LoadLoad, StoreStore, LoadStore and StoreLoad. To simplify the barrier semantics further
+WiredTiger defines acquire and release barriers, firstly we have the WT_ACQUIRE_READ_WITH_BARRIER
+macro which prevents LoadLoad and LoadStore reorderings. Next we have the
+WT_RELEASE_WRITE_WITH_BARRIER macro which prevents StoreStore and LoadStore reorderings, these
+barriers should be used as pairs. If one location requires an acquire barrier then there must be
+another location that requires a release barrier. Two more barrier constructs are defined in
+WiredTiger. WT_FULL_BARRIER which prevents all 4 kinds of described reorderings and
+WT_COMPILER_BARRIER which is a basic compiler barrier preventing compiler optimization from
+occurring across it, the compiler barrier does not impact CPU reordering.
```

Comment:
I suggest each WT barrier type should be enumerate into its own paragraph describing the purpose. I don't think a lot of detail is required.
Also potentially a link to the Preshing resource for further reference.

Reply:
[7c8a76d](https://github.com/wiredtiger/wiredtiger/pull/10575/commits/7c8a76d273ceb9d3513e47229079367559a7b4c8)